internal/controllers/machineset/machineset_controller.go

```diff
@@ -296,7 +302,12 @@ func (r *Reconciler) reconcile(ctx context.Context, cluster *clusterv1.Cluster,
 		filteredMachines = append(filteredMachines, machine)
 	}
 
+	result := ctrl.Result{}
 	// Remediate failed Machines by deleting them.
```

Comment:
I would change this section to the following:
```go
	var errs []error
	machinesToRemediate := make([]*clusterv1.Machine, 0, len(filteredMachines))
	for _, machine := range filteredMachines {
		// filteredMachines contains machines in deleting status to calculate correct status.
		// skip remediation for those in deleting status.
		if !machine.DeletionTimestamp.IsZero() {
			continue
		}
		if conditions.IsFalse(machine, clusterv1.MachineOwnerRemediatedCondition) {
			machinesToRemediate = append(machinesToRemediate, machine)
		}
	}

	if len(machinesToRemediate) > 0 {
		preFlightChecksResult, err := r.runPreFlightChecks(ctx, cluster, machineSet)
		if err != nil {
			return ctrl.Result{}, err
		}
		// Delete the machines only if the pre-flight checks have passed. Do not delete machines if we cannot
		// guarantee creating new machines.
		if preFlightChecksResult.IsZero() {
			for _, machine := range machinesToRemediate {
				log.Info("Deleting machine because marked as unhealthy by the MachineHealthCheck controller")
				patch := client.MergeFrom(machine.DeepCopy())
				if err := r.Client.Delete(ctx, machine); err != nil {
					errs = append(errs, errors.Wrap(err, "failed to delete"))
					continue
				}
				conditions.MarkTrue(machine, clusterv1.MachineOwnerRemediatedCondition)
				if err := r.Client.Status().Patch(ctx, machine, patch); err != nil && !apierrors.IsNotFound(err) {
					errs = append(errs, errors.Wrap(err, "failed to update status"))
				}
			}
		} else {
			result = util.LowestNonZeroResult(result, preFlightChecksResult)
		}
	}
```

Then we only run the preflight checks when we actually have to